### DISCUSSION TITLE
Missing WriteableBitmap.WritePixels?

### DISCUSSION THREAD

**Discussion by uno-studio** [AvaloniaUI/Avalonia#7661] [General]
I can't seem to find it WriteableBitmap.WritePixels function?

**Comment by maxkatz6** [comment]
Use ```WriteableBitmap.Lock``` method so you will get writeable [ILockedFramebuffer](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Visuals/Platform/ILockedFramebuffer.cs).

**Reply by uno-studio** [reply]
> Use `WriteableBitmap.Lock` method so you will get writeable [ILockedFramebuffer](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Visuals/Platform/ILockedFramebuffer.cs).

in wpf, i am using this code to output WriteableBitmap  with `ImageBgr24` input from Geb.Image library. For avalonia how would I convert the code to achieve the same?
```c#
        private ImageBgr24 imgCache;

        private WriteableBitmap bmpCache;

        public void OnRender(ImageBgr24 img)
        {
            if (img == null) return;
            if (imgCache == null) imgCache = img.Clone();
            lock (imgCache)
                imgCache.CloneFrom(img);
            VideoDisplay?.Dispatcher.InvokeAsync(Render);
        }

        private void Render()
        {
            if (bmpCache == null)
            {
                bmpCache = new WriteableBitmap(imgCache.Width, imgCache.Height, 96.0, 96.0, PixelFormats.Bgr24, null);
                VideoDisplay.RenderTarget.Source = bmpCache;
            }
            lock (imgCache)
                bmpCache.WritePixels(new Int32Rect(0, 0, imgCache.Width, imgCache.Height), imgCache.StartIntPtr, imgCache.ByteCount, imgCache.Stride);
        }
```

**Reply by maxkatz6** [reply]
Use an Address pointer to write your image into the writable bitmap. You might as well copy byte data from your imgCache.StartIntPtr if format matches.

**Reply by uno-studio** [reply]
I still can't succeed :(

**Reply by uno-studio** [reply]
@kekekeks can help me? I want to update WriteableBitmap from ImageBgr24's data. i used WriteableBitmap .WritePixels in WPF, but in Avalonia, I don't know how, no example found. ImageBgr24 has a method 'ToBitmap', which can save to stream and create Avalonia.Media.Imaging.Bitmap, however it is too slow and resource consuming.

**Reply by maxkatz6** [reply]
You can use this as a reference
https://github.com/Gillibald/Avalonia/blob/fdffe1b1242798cbac96cfec41794b160c1805fc/src/Avalonia.Visuals/Platform/LockedFramebufferExtensions.cs#L30

**Comment by Shadowblitz16** [comment]
There really needs to be a better imaging api for avalonia.
I mean come on!

Those extension methods are probably internal and they use completely different formats.
Write takes a Color while Read Takes a span

We have span why aren't we using that?
Do we really have to work with raw pointers?
